woocommerce/woocommerce-gateway-stripe
# Improve error handling when saving a new non-UPE payment method

Fixes #2632

## Changes proposed in this Pull Request:

According to the related issue, `WC_Stripe_API::get_payment_method()` returns null in some situations, and this unexpected response from Stripe isn't handled gracefully. 

This PR improves the error handling by failing gracefully and adding more specific entries to the error log when this happens. These entries could also help us pinpoint what causes this unexpected response.

Changes
- Bail out when the retrieved source object is empty or has errors, instead of [moving forward](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/7.4.0/includes/abstracts/abstract-wc-stripe-payment-gateway.php#L1169-L1173)
- Add detailed entries to the logs when something goes wrong when retrieving the payment method or attaching it to the customer.
- Check whether the payment method is reusable (using [WC_Stripe_Helper::is_reusable_payment_method()](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/class-wc-stripe-helper.php#L819-L830) ) before attaching it to the customer.

**Public interface changes**
- Before, the variable `$source_object` passed to the `'wc_stripe_add_payment_method_' . $payment_method . '_success'` action wasn't an actual source object when the source ID [was retrieved](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/7.4.0/includes/abstracts/abstract-wc-stripe-payment-gateway.php#L1175C2-L1175C2) from `$_POST['stripe_token']`. If it didn't trigger a fatal error along the way, that is. Now, the `$source_object` variable is always a source object.
- Before, `WC_Stripe_Payment_Gateway::add_payment_method()` [returned](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/7.4.0/includes/abstracts/abstract-wc-stripe-payment-gateway.php#L1187) `null` when there was an error, which isn't expected by the [calling](https://woocommerce.github.io/code-reference/files/woocommerce-includes-class-wc-form-handler.html#source-view.527) `WC_Form_Handler ::add_payment_method_action()`. Now, it returns an array with the `result => failure` key-value pair.
- The `WC_Stripe_Payment_Gateway::save_payment_method()` method was moved from [WC_Stripe_Payment_Gateway](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/7.4.0/includes/class-wc-gateway-stripe.php#L504-L524) to the abstract [WC_Stripe_Payment_Gateway](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/abstracts/abstract-wc-stripe-payment-gateway.php#L15). WC_Stripe_Payment_Gateway extends this abstract class so it'll continue to inherit it. But it'll also be available now for other classes extending the abstract WC_Stripe_Payment_Gateway class.

## Testing instructions

**Testing the problem**

1. Make sure UPE is disabled. Do this by going to Stripe settings `{site url}/wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe&panel=settings` -> Advanced settings -> Uncheck "Enable the updated checkout experience"
2. In this same section, enable "Log error messages" 
3. Change your store currency to Euros
4. Under the Payment methods settings page, `{site url}/wp-admin/admin.php?page=wc-settings&tab=checkout`, enable SEPA Direct Debit
5. Make the `get_source_object` [method](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/7.4.0/includes/abstracts/abstract-wc-stripe-payment-gateway.php#L643) return an empty object or null. `return (object) [];` / `return null;`
6. As a logged-in shopper, go to the "Add payment method page". `{url}/my-account/add-payment-method/`
7. Select SEPA and submit the form.
8. Confirm that:
  - You can't add the payment method
  - A WC error notice is displayed to the shopper saying `There was a problem adding the payment method. Unable to add payment method to your account.`
  - There are no fatal errors or warnings
  - There's an entry in the Stripe log (at `{site url}/wp-admin/admin.php?page=wc-status&tab=logs`) saying "The retrieved source doesn't contain an ID"
9. Remove the forced return value you added to `get_source_object` [method](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/7.4.0/includes/abstracts/abstract-wc-stripe-payment-gateway.php#L643) in step 5 to perform the regression tests.

**Regression tests**

Confirm you can still save payment methods through the checkout page
1. Make sure UPE is disabled
2. As a shopper, go to the "Add payment method" page. `{url}/my-account/add-payment-method/`
3. Add a CC. `5555555555554444` any expiration date, any CVC number
4. Confirm it gets saved
5. Add a SEPA payment method
6. Confirm it gets saved


Confirm you can save a new non-UPE payment method through the Checkout page. 
1. Make sure UPE is disabled
2. As a shopper, add a product to your cart and proceed to the checkout page
3. Select "Credit Card" and click on "Use a new payment method"
4. Check off " Save payment information to my account for future purchases"
5. Click on "Place order"
6. Go to the Payment Methods page `{site url}/my-account/payment-methods/`
7. Confirm the CC shows up there

Repeat these steps but use SEPA instead of Credit Card in step 3. 

Confirm 3DS cards get saved when paying for an order through the Pay for Order page
1. As an admin, create an order. You can use this same admin user as the order owner
2. Visit the Pay for Order page for this order
3. Select "Credit Card" and select "Use a new payment method"
4. Check off " Save payment information to my account for future purchases"
5. Enter a 3DS card. Like `4000000000003220`
5. Click on "Place order"
6. Go to the Payment Methods page `{site url}/my-account/payment-methods/`
7. Confirm the CC shows up there

---


**Post merge**
